perf(ngOptions): prevent initial options repaiting

pbr1111 opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.


**What is the current behavior? (You can also link to an open issue here)**
Refers to this issue: [ngOptions slow performance in IE due to option rerendering](https://github.com/angular/angular.js/issues/15801)


**What is the new behavior (if this is a feature change)?**
Now the options are loaded once when we create the `<select ng-options>` directive. Also, if we add an `<option>` element with an ng-if directive, it is not deleted now.


**Does this PR introduce a breaking change?**
No. 


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:
None.

Narretz commented:
Hi,
thanks for the Pr. As you can see in the Travis logs, there are a lot of test failures. Most of these come from the fact that there's a different writeSelectValue for select multiple, so you'll to make the change to this fn too.

Please also note that we definitely need to support an empty option with ngIf - there are some tests for this, and I assume some of them will fail in the current state.

You can run the unit tests locally with `grunt autotest`. This runs the test every time you change a file.

pbr1111 commented:
Hi, let's see if I can fix it. Thanks for the info!

pbr1111 commented:
I just uploaded the fixed code. Let's see if it pass the tests.

pbr1111 committed:
perf(ngOptions): prevent initial options repaiting

Avoid double execution of `updateOptions()` method, which causes a complete repainting of all `<option>` elements.

Fixes angular/angular.js#15801

pbr1111 commented:
The code just passed all the tests. This solution avoids double repainting of `<options>` and also makes ng-if work correctly in the empty option.

Narretz commented:
Thx @pbr1111, I will review this next week. Was there any reason for closing / reopening the issue so often?

pbr1111 commented:
I close and reopen because eslint tests fails several times and I've made a rebase of my branch each time it fails to upload the new commit with the errors solved.

Narretz commented:
If you push new / changed code, it'll run Travis again, no need to close / reopen the issue.

gkalpak commented on src/ng/directive/ngOptions.js:
```diff
@@ -624,7 +629,7 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
         // See https://github.com/angular/angular.js/issues/11314 for more info.
         // This is unfortunately untestable with unit / e2e tests
-        if (option.label !== element.label) {
+        if (option.label !== undefined) {
```
Why this change?

pbr1111 replied:
element.label always will be undefined (because we create a new empty option element each time), then it's useless to read element.label value if we know that its value is undefined.

Narretz replied:
You should use `isDefined`here

gkalpak commented on src/ng/directive/ngOptions.js:
```diff
@@ -420,12 +420,19 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
-          break;
-        }
+      var optionsToRemove = [];
+      for (var i = 0, children = selectElement[0].childNodes, ii = children.length; i < ii; i++) {
```
Is the move from jqLite/jQuery APIs to native APIs part of the perf fix or is it just a "generally good idea" type of change?

pbr1111 replied:
jqLite/jQuery children() method doesn't return the nodes of nodeType equal to NODE_TYPE_COMMENT. We can use jqLite/jQuery contents() method instead, but I prefer to use native API.

gkalpak commented on src/ng/directive/ngOptions.js:
```diff
@@ -420,12 +420,19 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
+      var optionsToRemove = [];
+      for (var i = 0, children = selectElement[0].childNodes, ii = children.length; i < ii; i++) {
+          if (!selectCtrl.hasEmptyOption
+              && (children[i].value === '' || children[i].nodeType === NODE_TYPE_COMMENT)) {
```
So, any comment will be interpreted as empty option? This doesn't seem right.

pbr1111 replied:
True. I'm doing changes to my commit. I will simplify this whole part and leave it as close to the previous code.

pbr1111 commented:
To modify the commit, is there any way to do it without having to rebase my branch?

pbr1111 commented:
I just modified the previous commit. Now the select element is emptied after it founds (or not) an empty option (the code is the same as before, it works because the options haven't yet been compiled in postLink, my previous solution was totally incorrect). The empty option element is readded if it's provided.

Narretz commented:
@pbr1111 You can just push a new commit and you / we will squash before merging.

> it works because the options haven't yet been compiled in postLink, my previous solution was totally incorrect

You should include a test that works with the correct version but fails with the incorrect one - it sounds like we are missing a test there

pbr1111 committed:
test(ngOptions): ignore comments within a <select> element

ngOptions should ignore comments within a `<select>`element and keep only the `<!-- ngIf -->` comment of the empty option.

Ref angular/angular.js#15812

pbr1111 committed:
test(ngOptions): ensure options aren't repainted when initialize ngOptions

Since https://github.com/angular/angular.js/commit/97b3e003bc60f979fddecd16bc50304b5bef7e53, ngOptions are repainting the options twice when it initializes. This test verifies that the options arent unnecessarily repainted.

Ref angular/angular.js#15812

pbr1111 commented:
I added some tests to check both cases. The second test https://github.com/angular/angular.js/pull/15812/commits/2ec5c07a95d1f6a90151048b49e6579e7cf36539 should fail with the current implementation, but it doesn't fail with https://github.com/angular/angular.js/pull/15812/commits/eb3661c6e4d66c76c3cbaed099b21719308dea6a commit.

Narretz commented on src/ng/directive/ngOptions.js:
```diff
@@ -449,36 +452,38 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
         selectCtrl.writeValue = function writeNgOptionsValue(value) {
-          var selectedOption = options.selectValueMap[selectElement.val()];
-          var option = options.getOptionFromViewValue(value);
+          if (options) {
```
Can you change to `if (!options) return`? This will reduce the whitespace changes.

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2641,6 +2658,23 @@ describe('ngOptions', function() {
+      ];
+      scope.isBlank = true;
+
+      createSingleSelect('<!-- test comment --><option ng-if="isBlank" value="">blank</option><!-- test comment 2 -->');
```
Line is too long, please split at around 100 chars

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2641,6 +2658,23 @@ describe('ngOptions', function() {
     }));
 
+
+    it('should ignore the comments within the select', function() {
```
There are no expectations regarding the comments you added. Should this test  that all comments that are not part of the empty option are removed?

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
   });
 
+
+  it('should render the options only one time', function() {
```
the **initial** options

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
+      'ng-model': 'value'
+    });
+
+    element.attr('ng-options', 'value for value in values');
```
you should pass the ngoptions expression like you did with ng-model. Or use createSingleSelect which has useful defaults.

Narretz commented on src/ng/directive/ngOptions.js:
```diff
@@ -428,6 +428,9 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
       }
 
+      // Remove all the elements inside the select.
+      selectElement.empty();
```
Please add something like:
"The empty option will be compiled and rendered before we first generate the options"

Narretz reviewed (changes requested):
Looking good so far! Some minor clarifications and changes nedded.  I would like for @petebacondarwin too also take a look because I remember he once wasn't a fan of doing `if (!options) return`

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -779,6 +779,23 @@ describe('ngOptions', function() {
+    var repaint = 0;
+    element[0].addEventListener('DOMSubtreeModified', function(ev) {
+      repaint++;
+      expect(repaint).toBeLessThan(2);
```
Firefox 47 failed this expect. Probably because `DOMSubtreeModified`behaves differently in it. Maybe we should use MutationObserver instead for Browsers that support it?

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -779,6 +779,32 @@ describe('ngOptions', function() {
+      'ng-model': 'value'
+    });
+
+    // Add ngOptions as attribute before recompiling the select element
```
Thanks for clarifying why you did it like this ;)
However, recompiling is generally a bad idea, and I wouldn't put it in a test. The better option is to add the listener via a custom directive.

Narretz commented:
Unfortunately, Firefox still reports 2 paints ... :(
